File: S/2012-sum-of-beauty-in-the-array/main.go
package main

import "fmt"

func sumOfBeauties(nums []int) int {
	lBuf, rBuf := make([]int, len(nums)), make([]int, len(nums))
	max := -1
	for i := 0; i < len(nums); i++ {
		if nums[i] > max {
			max = nums[i]
		}
		lBuf[i] = max
	}
	min := 1000000
	for i := len(nums) - 1; i >= 0; i-- {
		if nums[i] < min {
			min = nums[i]
		}
		rBuf[i] = min
	}

	ret := 0
	for i := 0; i < len(nums); i++ {
		if i == 0 || i == len(nums)-1 {
			continue
		}
		if nums[i] > lBuf[i-1] && nums[i] < rBuf[i+1] {
			ret += 2
		} else if nums[i] > nums[i-1] && nums[i] < nums[i+1] {
			ret++
		}
	}
	return ret
}

func main() {
	fmt.Println()
}

File: S/2012-sum-of-beauty-in-the-array/main.rs
struct Solution {}

impl Solution {
  pub fn sum_of_beauties(nums: Vec<i32>) -> i32 {
    let mut left: Vec<i32> = vec![0; nums.len()];
    left[0] = nums[0];
    let mut right: Vec<i32> = vec![0; nums.len()];
    right[nums.len() - 1] = nums[nums.len() - 1];
    (1..nums.len() - 1).for_each(|idx| {
      left[idx] = left[idx - 1].max(nums[idx]);
      right[nums.len() - 1 - idx] = right[nums.len() - idx].min(nums[nums.len() - 1 - idx]);
    });

    (1..nums.len() - 1)
      .map(|idx| {
        match (
          nums[idx] > left[idx - 1] && nums[idx] < right[idx + 1],
          nums[idx] > nums[idx - 1] && nums[idx] < nums[idx + 1],
        ) {
          (true, _) => 2,
          (false, true) => 1,
          (_, _) => 0,
        }
      })
      .collect::<Vec<i32>>()
      .iter()
      .sum()
  }
}
